pkg/engine/forceMutate.go

```diff
@@ -0,0 +1,107 @@
+package engine
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+
+	kyverno "github.com/nirmata/kyverno/pkg/api/kyverno/v1"
+	"github.com/nirmata/kyverno/pkg/engine/context"
+	"github.com/nirmata/kyverno/pkg/engine/mutate"
+	"github.com/nirmata/kyverno/pkg/engine/response"
+	"github.com/nirmata/kyverno/pkg/engine/utils"
+	"github.com/nirmata/kyverno/pkg/engine/variables"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func mutateResourceWithOverlay(resource unstructured.Unstructured, overlay interface{}) (unstructured.Unstructured, error) {
+	patches, err := mutate.MutateResourceWithOverlay(resource.UnstructuredContent(), overlay)
+	if err != nil {
+		return unstructured.Unstructured{}, err
+	}
+	if len(patches) == 0 {
+		return resource, nil
+	}
+
+	// convert to RAW
+	resourceRaw, err := resource.MarshalJSON()
+	if err != nil {
+		return unstructured.Unstructured{}, err
+	}
+
+	var patchResource []byte
+	patchResource, err = utils.ApplyPatches(resourceRaw, patches)
+	if err != nil {
+		return unstructured.Unstructured{}, err
+	}
+
+	resource = unstructured.Unstructured{}
+	err = resource.UnmarshalJSON(patchResource)
+	if err != nil {
+		return unstructured.Unstructured{}, err
+	}
+
+	return resource, nil
+}
+
+// ForceMutate does not check any conditions, it simply mutates the given resource
+func ForceMutate(ctx context.EvalInterface, policy kyverno.ClusterPolicy, resource unstructured.Unstructured) (unstructured.Unstructured, error) {
+	var err error
+	for _, rule := range policy.Spec.Rules {
+		if !rule.HasMutate() {
+			continue
+		}
+
+		mutation := rule.Mutation.DeepCopy()
+
+		if mutation.Overlay != nil {
+			overlay := mutation.Overlay
+			if ctx != nil {
+				if overlay, err = variables.SubstituteVars(ctx, overlay); err != nil {
```

Comment:
This `ForceMutate` function is only called once and we pass nil to ctx. Do we need to substitute variables here if ctx != nil?

Reply:
In the current implementation ctx is never passed in - i added this code in the hopes that this can more easily be able to become a subset for engine.Mutate.

Policy validation will never need to send ctx so it will not have the above issues.

When it eventually becomes part of engine.mutate - the ctx will be valid since its generated when resource is being applied to policy